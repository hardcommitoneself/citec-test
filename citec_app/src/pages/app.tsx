// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useEffect } from 'react'
import { useLocation, useRoutes } from 'react-router-dom'
import { Loading } from 'features/ui'
import { routes } from 'routes'
import { useAppReduxLoader, useGetAuth0Token, useAppSubscriptions } from 'features/utils/hooks'

import TagManager from 'react-gtm-module'

function App() {
   const { pathname } = useLocation()

   const { isAuthenticated, isLoading, loginWithRedirect, error, logout, token, user } =
      useGetAuth0Token()

   const location = useLocation()

   useAppReduxLoader(token)

   useEffect(() => {
      if (error?.message === 'Access denied.') {
         logout()
      }

   }, [])

   useAppSubscriptions(user, pathname)

   useEffect(() => {
      const tagManagerArgs = {
         gtmId: process.env.REACT_APP_GOOGLE_TAG_MANAGER_ID as string
      }

      TagManager.initialize(tagManagerArgs)

   }, [location])

   const element = useRoutes(routes)

   if (isLoading) return <Loading />

   return (
      <div className="App" data-testid="app-container">
         {isAuthenticated ? element : loginWithRedirect()}
      </div>
   )
}

export default App