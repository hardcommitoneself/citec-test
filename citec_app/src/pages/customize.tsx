// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { Accordion, AccordionDetails, AccordionSummary } from '@mui/material'
import { useEffect } from 'react'
import { useDispatch } from 'react-redux'

import { setSectors } from 'store/slices/universe'
import { useAppSelector } from 'store/hooks'
import { useGetCustomUniverse } from 'features/utils/hooks/use-get-custom-universe'
import { Icon } from 'assets/icons'
import {
   Factors,
   SectorsUniverse,
   Tickers,
   TotalsUniverse,
   UniverseSelection,
} from 'features/customize'

export const Customize = () => {
   const customUniverse = useGetCustomUniverse()

   const dispatch = useDispatch()

   const sectors = useAppSelector((state) => state.static_data.universe.sectors)
   const sectorsFilter = useAppSelector((state) => state.universe.sectors_filter)

   const parameters = [
      { name: 'Market Screener', component: <UniverseSelection /> },
      { name: 'Sector Screener', component: <SectorsUniverse /> },
      { name: 'Smart Beta Factor Screener', component: <Factors /> },
      {
         name: 'Include/Exclude',
         component: (
            <div className="grid grid-cols-2 gap-3">
               <Tickers type="Include" /> <Tickers type="Exclude" />
            </div>
         ),
      },
   ]

   const parametersWithInfo = ['Sector Screener', 'Smart Beta Factor Screener']

   useEffect(() => {
      if (sectorsFilter.length === 0) {
         dispatch(setSectors(sectors))
      }
   }, [])

   return (
      <>
         <TotalsUniverse totals={customUniverse} />

         <div className="my-3">
            {parameters.map(({ name, component }, i) => (
               <Accordion key={i}>
                  <AccordionSummary>
                     <p className="flex items-center text-xl my-2 text-[#2d405a]">
                        <span className="font-semibold mr-2 ">{name}</span>
                        {parametersWithInfo.includes(name) && <Icon.Info />}
                     </p>
                  </AccordionSummary>
                  <AccordionDetails>{component}</AccordionDetails>
               </Accordion>
            ))}
         </div>
      </>
   )
}
