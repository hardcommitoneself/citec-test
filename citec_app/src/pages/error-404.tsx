// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useNavigate } from 'react-router-dom'

export const Error404 = () => {
   const navigate = useNavigate()

   return (
      <div className="z-[60] absolute left-0 top-0 bg-white w-full h-full flex items-center justify-center">
         <div className="flex flex-col items-center">
            <div className="flex items-center justify-center">
               <h4 className="-rotate-90 text-red-600 text-2xl uppercase font-bold">
                  error
               </h4>

               <h2 className="text-9xl font-bold text-red-600 ">404</h2>
            </div>
            <span className="my-3 text-lg ">
               The page you are trying to find does not exist
            </span>
            <button
               className="btn-footer mt-2 w-full"
               onClick={() => navigate('/')}
            >
               Go home
            </button>
         </div>
      </div>
   )
}
