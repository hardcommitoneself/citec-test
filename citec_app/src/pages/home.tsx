// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useEffect } from 'react'
import { Button } from '@mui/material'
import { useAuth0 } from '@auth0/auth0-react'
import { useDispatch } from 'react-redux'

import { AppLink } from 'routes/AppLink'
import { AlpacaBalance } from 'features/brokers/alpaca/ui/components/alpaca-balance'
import { AlpacaBuyingPower } from 'features/brokers/alpaca/ui/components/alpaca-buying-power'
import { AlpacaMyInvest } from 'features/brokers/alpaca/ui/components/alpaca-my-invest'
import { InvestTable } from 'features/brokers/utils/table/invest-table'
import { LinkAccount } from 'features/brokers/link-account'
import { PageContainer } from 'features/ui/page-container'
import { renderColumnIdeas } from 'features/ui/table/columns/my-portfolios'
import { resetPortfolio } from 'store/slices/portfolio'
import { resetPortfolioOptimized } from 'store/slices/portfolio-optimized'
import { resetRebalance } from 'store/slices/rebalance'
import { resetTaxHarvesting } from 'store/slices/tax-harvesting'
import { resetTrades } from 'store/slices/trades'
import { resetUniverse } from 'store/slices/universe'
import { resetWorkflow } from 'store/slices/workflow'
import { useAppSelector } from 'store/hooks'
import { useHandlerMyIdeasTable } from 'features/utils/hooks/use-handler-my-ideas-table'
import { useValidateRegisteredUser } from 'features/utils/hooks'
import { resetStaticData } from 'store/slices/static-data'

export interface MyPortfolio {
   portfolio_id: number
   investment: number
   portfolio: string
   benchmark: string
   creation_date: string
   columnName: string
}

export const Home = () => {
   const { user } = useAuth0()
   const dispatch = useDispatch()

   const brokers = useAppSelector((state) => state.brokers)

   const { dataIdeasTable, loadIdeaOnRedux, removeIdeaFromRedux } =
      useHandlerMyIdeasTable()

   useValidateRegisteredUser(user?.email as string)

   useEffect(() => {
      dispatch(resetStaticData())
      dispatch(resetPortfolio())
      dispatch(resetUniverse())
      dispatch(resetRebalance())
      dispatch(resetPortfolioOptimized())
      dispatch(resetWorkflow())
      dispatch(resetTrades())
      dispatch(resetTaxHarvesting())
   }, [brokers.brokerConnected])

   return (
      <div>
         {brokers.brokerConnected !== '' && (
            <>
               <AlpacaBalance />
               <AlpacaBuyingPower />
               <AlpacaMyInvest />
            </>
         )}
         <PageContainer
            titlePage="My ideas"
            bgWhite
            actionBtn={
               <>
                  {brokers.brokerConnected !== 'undefined' &&
                     brokers.brokerConnected === '' && <LinkAccount />}
                  <AppLink to="/workflow">
                     <Button variant="home">
                        <span className="text-sm">Create+</span>
                     </Button>
                  </AppLink>
               </>
            }
         >
            {!!dataIdeasTable && (
               <InvestTable
                  columns={renderColumnIdeas(loadIdeaOnRedux, removeIdeaFromRedux)}
                  data={dataIdeasTable}
                  messageNoData=""
               />
            )}
         </PageContainer>
      </div>
   )
}
