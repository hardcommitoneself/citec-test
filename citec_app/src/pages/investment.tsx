// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import { Button, Slider } from '@mui/material'
import { useEffect, useState } from 'react'
import NumberFormat from 'react-number-format'
import { PageContainer } from 'features/ui/page-container'
import { setCash, setUpdatedCash } from 'store/slices/portfolio'
import { renderPopup } from 'features/popups/investment'
import { useAppSelector } from 'store/hooks'
import {
   calculateProgressionSavingFeesAndTaxes,
   ChartSimulation,
   FormPercentageRebalance,
   PotentialsSavings,
} from 'features/investment'
import {
   useAppNavigate,
   useHandlerTaxHarvesting,
   useValidateNumberStocks,
   useValidateSoldPosition,
   useValidationPorfolioLoaded,
} from 'features/utils/hooks'

const Group = require('assets/img/group.png')

export const Investment = () => {
   const dispatch = useDispatch()
   const navigate = useAppNavigate()
   const { soldCashOnTaxharvesting } = useValidateSoldPosition()
   const { checkValidStocks } = useValidateNumberStocks()
   useHandlerTaxHarvesting()

   const [investmentAcumulative, setInvestmentAcumulative] = useState(1000000)
   const [serieInvestmentPerYearWithSavings, setSerieInvestmentPerYearWithSavings] =
      useState([])
   const [
      serieInvestmentPerYearWithoutSavings,
      setSerieInvestmentPerYearWithoutSavings,
   ] = useState([])
   const [savings, setSavings] = useState({
      fees: 200000,
      taxes: 400000,
   })
   const [investment, setInvestment] = useState(1000000)
   const [years, setYears] = useState(10)
   const workflow = useAppSelector((state) => state.workflow.type)
   const MAX_VAL = 10000000

   const { isPortfolioLoaded } = useValidationPorfolioLoaded()

   const portfolioTickers = useAppSelector((state) => state.portfolio.ticker)

   const indexCash = portfolioTickers.indexOf('$CASH')
   const isCashOnPortfolio = indexCash >= 0

   const setPortfolioCash = () => {
      const cash = {
         outstanding_balance: investment,
         unrealized_capital_gains: 0,
      }

      dispatch(setUpdatedCash(cash))

      soldCashOnTaxharvesting(investment)
   }

   const investmentCash = useAppSelector(
      (state) => state.portfolio.outstanding_balance[indexCash]
   )

   const handleInvestmentChange = (value: number) => {
      setSavings({
         fees: calculateProgressionSavingFeesAndTaxes(
            years,
            'fees',
            investment,
            setInvestmentAcumulative,
            setSerieInvestmentPerYearWithSavings,
            setSerieInvestmentPerYearWithoutSavings
         ),
         taxes: calculateProgressionSavingFeesAndTaxes(
            years,
            'taxes',
            investment,
            setInvestmentAcumulative,
            setSerieInvestmentPerYearWithSavings,
            setSerieInvestmentPerYearWithoutSavings
         ),
      })
      setInvestment(value)
   }

   useEffect(() => {
      if (workflow === 'new') {
         handleInvestmentChange(investment)
      }
   }, [years])

   useEffect(() => {
      if (isCashOnPortfolio) {
         setInvestment(investmentCash)
      } else {
         dispatch(
            setCash({
               outstanding_balance: investment,
               unrealized_capital_gains: 0,
            })
         )
      }
      if (!isPortfolioLoaded) setPortfolioCash()

      soldCashOnTaxharvesting(isCashOnPortfolio ? investmentCash : investment)
   }, [])

   useEffect(() => {
      checkValidStocks()
   }, [investmentCash])

   return (
      <PageContainer>
         <div className="dashboard_wrapper">
            <div className="InvestmentBody dashboard_body">
               <div className="top_boxes_area_investment">
                  <div className={workflow == 'new' ? 'box' : 'box100'}>
                     {workflow == 'new' ? (
                        <h2>How much money you want to invest? </h2>
                     ) : (
                        <h2>Do you want to add more cash? </h2>
                     )}
                     <div className="inner_box">
                        <div className="top_area text-[#0BACA9]">
                           <h1>Investment in cash</h1>
                           <img
                              className="moreinfo"
                              src={Group}
                              alt=""
                              onClick={() => {
                                 renderPopup('investment')
                              }}
                              style={{ marginLeft: '3px' }}
                           />
                        </div>
                        <NumberFormat
                           value={investment}
                           thousandSeparator={true}
                           min={0}
                           maxLength={11}
                           isAllowed={({ value }) => {
                              if (Number(value) <= MAX_VAL) return true
                              return false
                           }}
                           onValueChange={({ value }) => {
                              handleInvestmentChange(Number(value))
                           }}
                           onBlur={setPortfolioCash}
                           prefix="$"
                           className="price_area w-full"
                           id="investment-input"
                        />

                        <Slider
                           valueLabelDisplay="auto"
                           onChange={(_, newValue) => {
                              setInvestment(newValue as number)
                           }}
                           onChangeCommitted={setPortfolioCash}
                           value={investment}
                           min={0}
                           max={10000000}
                        />
                     </div>
                  </div>
                  {workflow === 'new' && (
                     <PotentialsSavings
                        {...savings}
                        years={years}
                        serieInvestmentPerYearWithSavings={
                           serieInvestmentPerYearWithSavings
                        }
                        serieInvestmentPerYearWithoutSavings={
                           serieInvestmentPerYearWithoutSavings
                        }
                     />
                  )}
               </div>

               {workflow === 'new' ? (
                  <ChartSimulation
                     years={years}
                     setYears={setYears}
                     acumulative={investmentAcumulative}
                     serieInvestmentPerYearWithSavings={
                        serieInvestmentPerYearWithSavings
                     }
                     serieInvestmentPerYearWithoutSavings={
                        serieInvestmentPerYearWithoutSavings
                     }
                  />
               ) : (
                  <div className="box chart_box">
                     <h2 className="mb-[10px]">
                        How do you want to rebalance the portfolio?
                     </h2>
                     <FormPercentageRebalance />
                  </div>
               )}

               <div className="buttons_wrapper">
                  <Button
                     color="secondary"
                     onClick={() =>
                        navigate(
                           workflow === 'new' ? '/workflow' : '/portfolio-analysis'
                        )
                     }
                  >
                     Back
                  </Button>
                  <Button
                     onClick={() =>
                        navigate(
                           workflow === 'new' ? '/criteria' : '/tax-harvesting'
                        )
                     }
                  >
                     Next
                  </Button>
               </div>
            </div>
         </div>
      </PageContainer>
   )
}
