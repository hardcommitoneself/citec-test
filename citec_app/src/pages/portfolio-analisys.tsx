// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import { useEffect, useState } from 'react'
import Button from '@mui/material/Button'

import { BacktestRes } from 'features/utils/interfaces/backtest-api-resp'

import { Loading } from 'features/ui'
import { PageContainer } from 'features/ui/page-container'
import { PortfolioAnalisysResp } from 'features/utils/interfaces/portfolio-analysis-resp'
import { PositionsResp } from 'features/utils/interfaces/positions-resp'
import { setPortfolioAnalysis, setPortfolioPositions } from 'store/slices/portfolio'
import { useAppNavigate, useFetchApi } from 'features/utils/hooks'
import { useAppSelector } from 'store/hooks'
import {
   RiskreturnTab,
   Backtest,
   OverviewTab,
   Positions,
} from 'features/portfolio-analisys/tabs'
import { savePortfolio } from 'features/utils/save-portfolio'

export const PortfolioAnalysis = () => {
   const navigate = useAppNavigate()

   const { ticker, outstanding_balance, unrealized_capital_gains } = useAppSelector(
      (state) => state.portfolio
   )

   const outstanding_balance_total = outstanding_balance?.reduce(
      (accumulator, curr) => accumulator + curr
   )

   const benchmark = useAppSelector((state) => state.rebalance.benchmark)

   const { resp, isLoading, error } = useFetchApi<PortfolioAnalisysResp>({
      endpoint: '/ai/portfolio/analysis',
      method: 'POST',
      params: {
         hist_w: 5000,
      },
      payload: {
         portfolio: 'myportfolio',
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
      },
   })
   const { resp: backtest } = useFetchApi<BacktestRes>({
      endpoint: '/ai/portfolio/backtest',
      method: 'POST',
      params: {
         hist_w: 365,
      },
      payload: {
         portfolio: 'myportfolio',
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
      },
   })

   const { resp: positions } = useFetchApi<PositionsResp>({
      endpoint: '/ai/portfolio/positions',
      method: 'POST',
      payload: {
         portfolio: 'myportfolio',
         benchmark,
         ticker,
         outstanding_balance,
         unrealized_capital_gains,
      },
   })

   const handleSavePortfolio = () => {
      savePortfolio({
         navigate,
         benchmark,
         outstanding_balance,
         ticker,
         unrealized_capital_gains,
      })
   }

   const [active, setActive] = useState('Positions')
   const dispatch = useDispatch()

   const views = ['Positions', 'Overview', 'Risk', 'Backtest']

   useEffect(() => {
      if (resp) {
         dispatch(setPortfolioAnalysis({ ...resp }))
      }
      if (positions) {
         const newArr1 = positions.map((v) => ({
            ...v,
            weight: (
               (v.outstanding_balance * 100) /
               outstanding_balance_total
            ).toFixed(2),
         }))

         dispatch(setPortfolioPositions(newArr1))
      }
   }, [resp])

   const enabledViews = ['Positions', 'Risk', 'Overview', 'Backtest']

   if (isLoading) return <Loading />

   return (
      <PageContainer error={error}>
         <div className="InvestmentBody summary dashboard_body_portfolio_analysis">
            <div className="inner_dashboard_cri">
               <div className="options_summary_buttons">
                  {views.map((view) => (
                     <button
                        key={view}
                        className={active === view ? 'active-btn' : 'null'}
                        disabled={
                           //disabled is temporal, when the functionality is ready must remove this line
                           !enabledViews.includes(view)
                        }
                        onClick={() => {
                           setActive(view)
                        }}
                        data-testid={`btn-${view.toLowerCase}`}
                     >
                        {view}
                     </button>
                  ))}
               </div>
               {resp && backtest && (
                  <>
                     {active === 'Positions' && <Positions backtest={backtest} />}
                     {active === 'Backtest' && <Backtest backtest={backtest} />}
                     {active === 'Risk' && <RiskreturnTab />}
                     {active === 'Overview' && <OverviewTab backtest={backtest} />}
                  </>
               )}
               <div className="buttons_wrapper">
                  <Button
                     className="w-28 h-12"
                     color="secondary"
                     onClick={() => navigate('goBack')}
                  >
                     <span className="text-base font-normal">Back</span>
                  </Button>

                  <Button
                     className="w-28 h-12"
                     variant="save"
                     onClick={handleSavePortfolio}
                  >
                     <span>Save</span>
                  </Button>
               </div>
            </div>
         </div>
      </PageContainer>
   )
}
