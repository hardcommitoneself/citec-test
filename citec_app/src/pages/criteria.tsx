// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { Slider } from '@mui/material'
import { useDispatch } from 'react-redux'
import { useEffect, useState } from 'react'
import Button from '@mui/material/Button'
import Swal from 'sweetalert2'
import { Icon } from 'assets/icons'

import { PageContainer } from 'features/ui/page-container'
import { setNumberStocks, setObjective } from 'store/slices/rebalance'
import { useAppNavigate, useValidateNumberStocks } from 'features/utils/hooks'
import { useAppSelector } from 'store/hooks'
import { useSelectUniverse } from 'features/utils/hooks/use-select-universe'
import { ButtonAndModalCustomize } from 'features/customize'

const infoGray = require('assets/img/group.png')
const infoGreen = require('assets/img/group_green.png')

export const Criteria = () => {
   const dispatch = useDispatch()
   const navigate = useAppNavigate()

   const objectives = useAppSelector(
      (state) => state.static_data.rebalance.objectives
   )

   const benchmarks = useAppSelector(
      (state) => state.static_data.rebalance.benchmarks
   )
   const objective = useAppSelector((state) => state.rebalance.objective)
   const benchmark = useAppSelector((state) => state.rebalance.benchmark)
   const stocks = useAppSelector((state) => state.rebalance.number_stocks)
   const [rangeStocks, setRangeStocks] = useState(stocks)

   const { handleSelectBenckmark } = useSelectUniverse()
   const { isValidPortfolio } = useValidateNumberStocks()

   useEffect(() => {
      if (!isValidPortfolio)
         Swal.fire({
            icon: 'error',
            title: 'Oops...',
            text: 'The total amount of the portfolio is low. It is recommended to add more cash',
         })
   }, [isValidPortfolio])

   const benchmarkTitleAlert = () => {
      Swal.fire({
         html: `<div class="align-left">
            <b>The benchmark</b> is the market index against which you will be
            comparing your portfolio. Generally indexes are well-diversified and 
            a good representation of the overall market performance.
            </div>`,
         confirmButtonColor: '#0baca9',
         confirmButtonText: '<a href="about:blank" target="_blank">Learn more</a>',
         showCloseButton: true,
      })
   }

   const objectiveAlert = (key: any) => {
      switch (key) {
         case 'min_tracking_error':
            Swal.fire({
               html: `<div class="align-left">
                  <b>Direct indexing</b> consist of replicating the index/benchmark's
                  performance by selecting the combination of stocks and weights among 
                  a given universe that allows to minimize the tracking error.
                  </div>`,
               confirmButtonColor: '#0baca9',
               confirmButtonText:
                  '<a href="about:blank" target="_blank">Learn more</a>',
               showCloseButton: true,
            })
            break
         case 'min_neg_sharpe_ratio':
            Swal.fire({
               html: `<div class="align-left">
                  <b>The Sharpe ratio</b> is measured as return over volatility (Risk 
                  proxy). This optimization will look for the combination of stocks and 
                  weigths among a given universe that allows to maximize this ratio 
                  for simulated scenarios.
                  </div>`,
               confirmButtonColor: '#0baca9',
               confirmButtonText:
                  '<a href="about:blank" target="_blank">Learn more</a>',
               showCloseButton: true,
            })
            break
         case 'min_volatility':
            Swal.fire({
               html: `<div class="align-left">
                  <b>Volatility</b> is measured as the standard deviation of the 
                  portoflio's value. This optimization will look for the combination of
                  stocks and weigths among a given universe that allows to minimize 
                  the portfolio volatility for the simulated scenarios.
                  </div>`,
               confirmButtonColor: '#0baca9',
               confirmButtonText:
                  '<a href="about:blank" target="_blank">Learn more</a>',
               showCloseButton: true,
            })
            break
         case 'min_risk_parity':
            Swal.fire({
               html: `<div class="align-left">
                  <b>Risk parity</b> consists in allocating the same level of risk
                  to each of a universe's stocks. The more volatility, the less
                  capital you allocate to those stocks.
                  </div>`,
               confirmButtonColor: '#0baca9',
               confirmButtonText:
                  '<a href="about:blank" target="_blank">Learn more</a>',
               showCloseButton: true,
            })
            break
         default:
            break
      }
   }

   const stocksAlert = () => {
      Swal.fire({
         html: `<div class="align-left">
            <p>The number of positions should be less than investment/$300
            (average stock price). Additionally, the recommended number of 
            stocks differs depending on the type of optimization:</p>
            <ul style="list-style-type:square">
               <li>
               <b>-Direct Indexing:</b> +80 positions for tracking errors <0.5%
               </li>
               <li>
               <b>-Sharpe ratio:</b> +20 positions to leverage diversification
               </li>
               <li>
               <b>-Min volatility:</b> +20 postions to leverage diversification
               </li>
               <li>
               <b>-Risk parity:</b> N/A
               </li>
            </ul>
            </div>`,
         confirmButtonColor: '#0baca9',
         confirmButtonText: '<a href="about:blank" target="_blank">Learn more</a>',
         showCloseButton: true,
      })
   }

   const benchmarksFormatter = (benchmark: string) => {
      switch (benchmark) {
         case 'VTI':
            return 'Total US Market'
         case 'QQQ':
            return 'Nasdaq 100'
         case 'SPY':
            return 'S&P 500'
         case 'SNPE':
            return 'S&P 500 ESG'
         case 'IWM':
            return 'Russell 2000'
         default:
            break
      }
   }

   const objectivesFormatter = (objective: string) => {
      switch (objective) {
         case 'min_tracking_error':
            return 'Direct Indexing'
         case 'min_risk_parity':
            return 'Risk Parity'
         case 'min_volatility':
            return 'Min Volatility'
         case 'max_diversification_ratio':
            return 'Max Diversification'
         default:
            break
      }

      return ''
   }

   const filteredObjectives = objectives.filter(({ key }) => {
      const allowedObjectives = [
         'min_tracking_error',
         'min_volatility',
         'max_diversification_ratio',
         'min_risk_parity',
      ]

      return allowedObjectives.indexOf(key) > -1
   })

   const renamedObjectives = filteredObjectives.map((obj) => ({
      ...obj,
      name: objectivesFormatter(obj.key),
   }))

   const sortedObjectives = renamedObjectives.sort((a, b) => {
      //sort alphabetical

      return a.name.localeCompare(b.name)
   })

   const objectivesIcons = (objective: string) => {
      switch (objective) {
         case 'min_tracking_error':
            return <Icon.DirectIndexing />
         case 'min_risk_parity':
            return <Icon.RiskParity />
         case 'min_volatility':
            return <Icon.MinVolatility />
         case 'max_diversification_ratio':
            return <Icon.MaxDiversification />
         default:
            break
      }
   }

   return (
      <PageContainer
         titlePage="What do you want to optimize?"
         actionBtn={<ButtonAndModalCustomize />}
      >
         <div className="InvestmentBody dashboard_body">
            <div className="inner_dashboard_cri">
               <div className="select_wrapper">
                  <h1>
                     Select benchmark
                     <img
                        className="moreinfo"
                        src={infoGreen}
                        alt=""
                        onClick={() => {
                           benchmarkTitleAlert()
                        }}
                     />
                  </h1>
                  <div className="boxwrapper grid_4">
                     {benchmarks.map(({ description, key, ticker }) => (
                        <div
                           className="box"
                           key={key}
                           onClick={() => handleSelectBenckmark(ticker)}
                        >
                           <input
                              checked={benchmark === ticker}
                              type="radio"
                              name="benchmark"
                              id={description}
                              readOnly
                           />
                           <label>{benchmarksFormatter(key)}</label>
                        </div>
                     ))}
                  </div>
               </div>

               <div className="select_wrapper">
                  <h1>Select objective</h1>
                  <div className="boxwrapper grid_4">
                     {sortedObjectives.map(({ name, key }) => (
                        <div
                           className="box"
                           key={key}
                           onClick={() =>
                              dispatch(
                                 setObjective({
                                    name,
                                    key,
                                 })
                              )
                           }
                        >
                           <input
                              checked={objective.key === key}
                              type="radio"
                              name="objective"
                              id={name}
                              readOnly
                           />
                           <label htmlFor="DirectIndexing">
                              <div className="mr-2.5">{objectivesIcons(key)}</div>
                              {name}
                              {key != 'max_diversification_ratio' && (
                                 <img
                                    className="moreinfo"
                                    src={infoGray}
                                    alt=""
                                    onClick={() => {
                                       objectiveAlert(key)
                                    }}
                                 />
                              )}
                           </label>
                        </div>
                     ))}
                  </div>
               </div>

               <div className="select_wrapper">
                  <h1 style={{ marginBottom: 20 }}>
                     Select number of stocks
                     <img
                        className="moreinfo"
                        src={infoGreen}
                        alt=""
                        onClick={() => {
                           stocksAlert()
                        }}
                     />
                  </h1>

                  <Slider
                     value={rangeStocks}
                     onChangeCommitted={(_, newValue) =>
                        dispatch(setNumberStocks(newValue))
                     }
                     onChange={(_, newValue) => {
                        setRangeStocks(newValue as never[])
                     }}
                     valueLabelDisplay="on"
                     min={10}
                     max={200}
                  />
               </div>

               <div className="buttons_wrapper">
                  <Button
                     id="criteria_run_back_buttons"
                     color="secondary"
                     onClick={() => navigate('goBack')}
                  >
                     Back
                  </Button>
                  <Button
                     id="criteria_run_back_buttons"
                     variant="optimize"
                     onClick={() => navigate('/summary')}
                  >
                     Run
                  </Button>
               </div>
            </div>
         </div>
      </PageContainer>
   )
}
