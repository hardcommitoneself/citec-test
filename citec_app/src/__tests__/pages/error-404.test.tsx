// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { render } from '@testing-library/react'
import { Error404 } from 'pages'
import { BrowserRouter } from 'react-router-dom'

describe('test in error404 page', () => {
   test('Snapshot', () => {
      const { container } = render(
         <BrowserRouter>
            <Error404 />
         </BrowserRouter>
      )

      expect(container).toMatchSnapshot()
   })
})
