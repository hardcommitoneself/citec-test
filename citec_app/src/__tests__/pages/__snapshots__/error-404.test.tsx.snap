// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in error404 page Snapshot 1`] = `
<div>
  <div
    class="z-[60] absolute left-0 top-0 bg-white w-full h-full flex items-center justify-center"
  >
    <div
      class="flex flex-col items-center"
    >
      <div
        class="flex items-center justify-center"
      >
        <h4
          class="-rotate-90 text-red-600 text-2xl uppercase font-bold"
        >
          error
        </h4>
        <h2
          class="text-9xl font-bold text-red-600 "
        >
          404
        </h2>
      </div>
      <span
        class="my-3 text-lg "
      >
        The page you are trying to find does not exist
      </span>
      <button
        class="btn-footer mt-2 w-full"
      >
        Go home
      </button>
    </div>
  </div>
</div>
`;
